compiler/infra/ILTraverser.hpp

```diff
@@ -0,0 +1,173 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2017, 2017
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+#ifndef TR_ILTRAVERSER_HPP
+#define TR_ILTRAVERSER_HPP
+
+#include "infra/Checklist.hpp"
+#include "env/TRMemory.hpp"
+#include "compile/Compilation.hpp"
+
+#include <vector>
+#include <algorithm>
+
+namespace TR {
+
+// forward declaration of "imported" classes
+class ResolvedMethodSymbol;
+class Block;
+class TreeTop;
+class Node;
+
+/**
+ * @brief A class that abstracts aways the details of traversing Testarossa IL.
+ *
+ * This class uses the Observer pattern to separate the actual traversal mechanism
+ * from the code that needs to do the traversal. An observer can register itself
+ * with an instance of this class to be notified of the various transitions during
+ * the traversal.
+ *
+ * This class expects the input IL to be "sound"; properly structure for traversal.
+ */
+class ILTraverser
+   {
+   public:
+   explicit ILTraverser(TR::Compilation* comp);
+
+   /**
+    * @brief The base class for observers of an instance of ILTraverser
+    *
+    * Objects inheriting from this class can be registered with an instance of
+    * ILTraverser to be notified of the various transitions done during traversal.
+    *
+    * All signals have an empty body by default so that derived classes need
+    * only implement the signals they are interested in.
+    *
+    * All the events that can be listened to have the same general interface.
+    * They all take as argument a pointer to the object being visited. For each
+    * type of object that can be visited, there is one singal for when the object
+    * is entered, and one signal for when traversal returns to the object after
+    * having visited the object's children.
+    *
+    *    parent      --,
+    *       node    <--'  visitingNode(node)
+    *          child
+    *
+    *    parent
+    *       node        <--,  returnedToNode(node)
+    *          child     --'
```

Comment:
This is one case where perhaps the picture is less clear than the text (in my opinion).  In particular, when the function is actually called is not clear to me.  I'd say remove the graphics.